src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.cs

```diff
@@ -122,24 +134,36 @@ private static string GenerateAnonymousTypeName(int current)
                         {
                             return 0;
                         }
-                    }).ToList();
+                    }).ToImmutableArray();
             }
 
-            return transitiveAnonymousTypeReferences;
+            return structuralTypes;
         }
 
-        private static IList<INamedTypeSymbol> GetTransitiveNormalAnonymousTypeReferences(
-            ISet<INamedTypeSymbol> anonymousTypeReferences)
+        private static ImmutableArray<INamedTypeSymbol> GetTransitiveStructuralTypeReferences(
+            ImmutableArray<INamedTypeSymbol> structuralTypes)
         {
-            var transitiveReferences = new List<INamedTypeSymbol>();
-            var visitor = new NormalAnonymousTypeCollectorVisitor(transitiveReferences);
+            var transitiveReferences = new Dictionary<INamedTypeSymbol, (int order, int count)>();
+            var visitor = new StructuralTypeCollectorVisitor(transitiveReferences);
 
-            foreach (var type in anonymousTypeReferences)
-            {
+            foreach (var type in structuralTypes)
                 type.Accept(visitor);
+
+            using var _ = ArrayBuilder<INamedTypeSymbol>.GetInstance(out var result);
+
+            foreach (var (namedType, (order, count)) in transitiveReferences.OrderBy(kvp => kvp.Value.order))
+            {
+                if (namedType.IsTupleType && count == 1)
+                {
+                    // Ignore tuples only referenced once.  We'll keep them inline.  If the tuple shows up multiple times,
```

Comment:
I wonder if this should only apply if there isn't otherwise going to be a "Structural Types" section? eg, what happens in this case:

```
public (int a, int b) Goo((int a, int b), (string x, string y);
```

You're going to end up with a "Structural Types" section that doesn't list out all of the structural types.

Reply:
That's an interesting point. I could revise to make it that once we have at least one duplicated tuple that all tuples then are broken out...  That seems reasonable to me.